src/ladds/distributedMemParallelization/RankMigration.cpp

```diff
@@ -0,0 +1,198 @@
+/**
+ * @file RankMigration.cpp
+ * @author P. Gomez
+ * @date 2022-07-18
+ */
+#include "RankMigration.h"
+
+#include "ladds/distributedMemParallelization/ParticleCommunicator.h"
+#include "ladds/particle/Particle.h"
+
+std::vector<LADDS::Particle> LADDS::RankMigration::communicateParticles(std::vector<LADDS::Particle> &leavingParticles,
+                                                                        autopas::AutoPas<LADDS::Particle> &autopas,
+                                                                        const DomainDecomposition &decomposition) {
+  // Set up the communicator
+  const auto &comm = decomposition.getCommunicator();
+  const auto &localBoxMin = autopas.getBoxMin();
+  const auto &localBoxMax = autopas.getBoxMax();
+  auto logger = spdlog::get(LADDS_SPD_LOGGER_NAME);
+
+  ParticleCommunicator particleCommunicator;
+  std::vector<LADDS::Particle> incomingParticles;
+
+  auto getNeighborRank = [&](const auto &coordsThis, int direction, auto op) {
+    auto coordsOther = coordsThis;
+    coordsOther[direction] = op(coordsOther[direction], 1);
+    int rankOther{};
+    autopas::AutoPas_MPI_Cart_rank(comm, coordsOther.data(), &rankOther);
+    return rankOther;
+  };
+
+  if (const auto *regularGridDecomp = dynamic_cast<const RegularGridDecomposition *>(&decomposition)) {
```

Comment:
```suggestion
  // one case per domain decomposition
  if (const auto *regularGridDecomp = dynamic_cast<const RegularGridDecomposition *>(&decomposition)) {
```
Actually, this feels ugly. Maybe implement a function that basically is the body of this if case in every decomposition, call it on the decomposition object and let polymorphism figure out which one to actually use?

Reply:
You mean replace `dynamic_cast<const RegularGridDecomposition *>(&decomposition)` with `decomposition.getDecompositionPointer()` or so? Can't replace the assignment as we need the variable

Reply:
no, what I mean is:
* Create a new virtual function in `DomainDecomposition` 
* Move the stuff in the respective if branches into overrides of this function of the respective decompositions
* In RankMigration just call this (or get rid of the RankMigration wrapper for this) and let the language choose the right implementation through the polymorphism logic

Reply:
Done in https://github.com/esa/LADDS/pull/144/commits/8d1924c36f160d12e3621476e402ab8001f197aa